### DISCUSSION TITLE
Writing a file is very slow when lsp is processing

### DISCUSSION THREAD

**Discussion by martinetd** [helix-editor/helix#7427] [Troubleshooting] [answered]
Trying helix again on a rust project here, it looks like writes (as triggered by `:w`) get delayed until the lsp is done?

So I often get to wait 20ish seconds while the file isn't written, as the said project takes horribly long to build (future hell?).

I couldn't get any log despite running with `hx -vvv`, but the status bar is nice enough to report it's abnormal with a line such as this:
```
Async job failed: request 18 timed out
```

Ultimately the problem is that the lsp takes that long; rust-analyser is a resource hog (memory is a problem as well when opening multiple helix instances... See https://github.com/helix-editor/helix/discussions/6533 where I asked if it'd be possible to not start it until first edit or something, but I guess just using another editor for viewing works well enough for now); it's just slow to get status back everytime I modify something, and not being able to write files timely makes it work.

I'm not sure how it works internally but perhaps file write could have priority?  at least :q refuses to quit while waiting for the save.
(I guess it'll probably be slightly less annoying once I use buffers more effectively and won't need to save as often, but even just delaying running 'cargo test' by 20s everytime is painful)

FTR, hx --version `helix 23.05 (7f5940be)`

Thanks for helix anyway! There are some nice ideas in here.

**Comment by the-mikedavis** [comment, marked as answer]
It can take a while to write if the language server is slow because the write waits on the document formatting response from the server. You can use `:toggle-option auto-format` to disable auto-formatting or disable it specifically for Rust:

```toml
# ~/.config/helix/languages.toml or <project dir>/.helix/languages.toml
[[language]]
name = "rust"
auto-format = false
```

**Reply by martinetd** [reply]
Thanks! I actually couldn't get the default `:format` to actually do anything in my environment so I wrongly assumed it wasn't configured (probably something in my setup), but waiting for formatting response makes sense if it's enabled.

Spawning another command instead of relying on lsp for formatting seems to also work, so I'll give using an external command as formatter a try:

```
[[language]]
name = "rust"
formatter = { command = "rustfmt" }
```

EDIT: grmbl, wrongly assumed again that rustfmt was a bit hard to use with stdin but this apparently ended up working without doing anything when not trying hard to process stdin with options :|
